os/pm/pm_timer/pm_timer_set.c

```diff
@@ -0,0 +1,198 @@
+
+        /* Case where there are no timers in the list */
+
+        if (g_pmTimer_activeList.head == NULL) {
+                sq_addlast((FAR sq_entry_t *)timer, &g_pmTimer_activeList);
+                pmdbg("timer added in the head of linked list\n");
+        }
+        
+        /* We should add the timer in the sorted position of delay time */
+
+         else {
+                pm_wakeup_timer_t *curr;
+                pm_wakeup_timer_t *prev;
+                prev = curr = (pm_wakeup_timer_t *)g_pmTimer_activeList.head;
+                unsigned int now = 0;
+
+                /* Advance to positive time */
+                while ((now += curr->delay) < 0 && curr->next) {
+                        prev = curr;
+                        curr = curr->next;
+                }
+
+                /* Advance past shorter delays */
+                while (now <= timer->delay && curr->next) {
+                        prev = curr;
+                        curr = curr->next;
+                        now += curr->delay;
+                }
+
+                /* timer should be added before the curr */
+                if (timer->delay < now) {
+                        
+                        timer->delay -= (now - curr->delay);
+                        curr->delay -= timer->delay;
+                        if (curr == (pm_wakeup_timer_t *)g_pmTimer_activeList.head) {
+                                sq_addfirst((FAR sq_entry_t *)timer, &g_pmTimer_activeList);
+                        } else {
+                                sq_addafter((FAR sq_entry_t *)prev, (FAR sq_entry_t *)timer, &g_pmTimer_activeList);
+                        }  
+
+                /* timer expire time is greater than every other timer. 
+                 * timer should be added to the end of the linked list*/
+                } else {
+                        timer->delay -= now;
+                        if (!curr->next) {
+                                sq_addlast((FAR sq_entry_t *)timer, &g_pmTimer_activeList);
+                        } else {
+                                sq_addafter((FAR sq_entry_t *)curr, (FAR sq_entry_t *)timer, &g_pmTimer_activeList);
+                        }
+                }
+                pmdbg("timer added in linked list by using while loop\n");
+        }
+}
+
+/************************************************************************
+ * Public Functions
+ ************************************************************************/
+
+/************************************************************************
+ * Name: pm_timer_set
+ *
+ * Description:
+ *   This function adds a wakeup timer in the g_pmTimer_activeList. So that it will be
+ *   invoked just before sleep when needed. 
+ * 
+ * Parameters:
+ *   timer_interval - expected board sleep duration
+ *
+ * Return Value:
+ *   0 - success
+ *   -1 - error
+ *
+ ************************************************************************/
+
+int pm_timer_set(unsigned int timer_interval)
```

Comment:
There are 2 API, pm_timer_set and pm_set_timer. Although both of them serve different purpose, but the naming is too hard to differentiate. Would you to change one of these?
For pm_set_timer, it is used to set the HW timer to wakeup, so I think it is feasible to named as such.
For pm_timer_set here, actually it is adding a virtual timer to the pm_timer pool, also the core logic seems to be the function above (ie. add_pm_timer, no point create a helper function to do that since there is only one caller, create another function incur additional overhead for branching, so you can just inline it). Thus, I suggest to merge the api above to here. And perhaps just rename to pm_timer_add (ie. both API and filename).